woocommerce/woocommerce
# Add framework to write more robust action scheduler based migrations.

### All Submissions:






### Changes proposed in this Pull Request:

From time to time, we have had the need to migrate a lot of data, either to power features or to fix bugs. However, we don’t have a shared mechanism to do this, so every developer working on implementing any kind of migration would do it from scratch. Since the developer implements it from scratch, we often also push bugs specific to the migration routine itself that we have previously solved before.

In this PR, I propose implementing a lightweight framework to write a migration that we can use most of the time. This framework will consider meta aspects of migration such as scheduling, re-entring, pausing, resuming, etc. We, anyway, need to work on [making migrations more robust](https://github.com/woocommerce/woocommerce/issues/32922) as part of the Custom order tables project, so it’s a good time to make it re-usable as we implement it.

We can create two classes – `BatchProcessingController` and`BatchProcessor`, where:

1. `BatchProcessor` will be an abstract class that will provide a framework for implementing details specific to migration. Such as methods to process batches. All the meta details like starting the process, error handling, updating progress status, etc., will be already implemented.
1. `BatchProcessingController` class will take care of running the migration as needed, starting again if it gets stuck, scheduling in the action scheduler queue, etc. This also implements a periodic check for all pending migration and schedules action for error’ed migrations as well.

In other words, we divide actual migration logic and logic for scheduling migration into two separate concerns. While the actual migration logic changes on case to case basis, the logic for scheduling the migrations can remain the same.

This PR also modifies custom order table migrations to use this common framework.



Closes #32922 .

### How to test the changes in this Pull Request:

The easiest way to test is to run the COT migration from the interface -

1. Go to WooCommerce > Status > Tools and delete and recreate custom order tables.
1. Go to WooCommerce > Settings > Advanced > Custom Data Stores and start the migration process.
1. Verify that migration starts and completes as expected.

### Other information:




### FOR PR REVIEWER ONLY:
